internal/app/agent/service.go

```diff
@@ -168,41 +178,78 @@ func (a *agent) Control(uuid, cmdStr string) error {
 		return err
 	}
 
-	payload, err := encodeSenML(uuid, cmd, resp)
-	if err != nil {
-		return err
-	}
-
-	if err := a.Publish(a.config.Agent.Channels.Control, string(payload)); err != nil {
-		return err
-	}
-
-	return nil
+	return a.processResponse(uuid, cmd, resp)
 }
 
 // Message for this command
-// "[{"bn":"1:", "n":"config", "vs":"export, /configs/export/config.toml, config_file_content"}]"
+// [{"bn":"1:", "n":"services", "vs":"view"}]
+// [{"bn":"1:", "n":"config", "vs":"save, export, filename, filecontent"}]
 // config_file_content is base64 encoded marshaled structure representing service conf
 // Example of creation:
 // 	b, _ := toml.Marshal(cfg)
 // 	config_file_content := base64.StdEncoding.EncodeToString(b)
 func (a *agent) ServiceConfig(uuid, cmdStr string) error {
 	cmdArgs := strings.Split(strings.Replace(cmdStr, " ", "", -1), ",")
-	if len(cmdArgs) < 3 {
+	if len(cmdArgs) < 1 {
 		return errInvalidCommand
 	}
+	resp := ""
+	cmd := cmdType(cmdArgs[0])
+
+	switch cmd {
+	case view:
+		services, err := json.Marshal(a.Services())
+		if err != nil {
+			return err
+		}
+		resp = string(services)
+	case save:
+		if len(cmdArgs) < 4 {
+			return errInvalidCommand
+		}
+		service := cmdArgs[1]
+		fileName := cmdArgs[2]
+		fileCont := cmdArgs[3]
+		if err := a.saveConfig(service, fileName, fileCont); err != nil {
+			return err
+		}
+	}
+	return a.processResponse(uuid, string(cmd), resp)
+}
 
-	service := cmdArgs[0]
-	fileName := cmdArgs[1]
-	fileCont, err := base64.StdEncoding.DecodeString(cmdArgs[2])
+func (a *agent) processResponse(uuid, cmd, resp string) error {
+	payload, err := encodeSenML(uuid, cmd, resp)
 	if err != nil {
 		return err
 	}
-	c := &export.Config{}
+	if err := a.Publish(a.config.Agent.Channels.Control, string(payload)); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (a *agent) saveConfig(service, fileName, fileCont string) error {
+	switch service {
+	case "export":
+		var content []byte
```

Comment:
This line is not necessary if you are defining content below.

Reply:
right